packages/alfa-rules/src/sia-er62/rule.ts

```diff
@@ -69,60 +61,128 @@ export default Rule.Atomic.of<Page, Element>({
+          if (isTextNode && container.isSome()) {
+            // For each link, store the parent of the text nodes it contains
+            if (link.isSome()) {
+              linkText = linkText.set(
+                link.get(),
+                linkText
+                  .get(link.get())
+                  .getOr(Set.empty<Element>())
+                  .add(node.parent().filter(isElement).get())
+              );
+            }
+
+            // For each container, store the parent of the text nodes it contains
+            if (link.isNone()) {
+              nonLinkText = nonLinkText.set(
+                container.get(),
+                nonLinkText
+                  .get(container.get())
+                  .getOr(Set.empty<Element>())
+                  .add(node.parent().filter(isElement).get())
+              );
+            }
+          }
+
           const children = node.children({
             flattened: true,
             nested: true,
           });
 
           for (const child of children) {
-            yield* visit(child, container);
+            gather(child, container, link);
+          }
+        }
+
+        function* getApplicableLinks(): Iterable<Element> {
+          // Check if foreground is the same with the parent <p> element
+          const hasDifferentForeground = (
+            link: Element,
+            container: Element
+          ): boolean =>
+            getForeground(link, device).some((linkColors) => {
+              // If the link has a foreground with the alpha channel less than 1 and background gradient color
+              // then the rule is applicable as we can't tell for sure if it ever has the same foreground with a link
+              // that might have the same foreground and gradien background, but with different gradient type or spread
+              if (linkColors.length !== 1) {
+                return true;
+              }
+              return getForeground(container, device).some(
+                (containerColors) => {
+                  // Similalry to the link, we assume it's applicable if the container has different foreground colors
+                  if (containerColors.length !== 1) {
+                    return true;
+                  }
+                  return !linkColors[0].equals(containerColors[0]);
+                }
+              );
+            });
+
+          for (const link of linkText.keys()) {
+            const linkTexts = linkText.get(link).get();
+            const nonLinkTexts = nonLinkText
+              .get(containers.get(link).get())
+              .get();
+
+            for (const linkElement of linkTexts) {
+              for (const nonLinkElement of nonLinkTexts) {
+                if (hasDifferentForeground(linkElement, nonLinkElement)) {
+                  applicableContainers = applicableContainers.set(
+                    link,
+                    nonLinkElement
+                  );
```

Comment:
It seems if we include the extra values, then in some cases we will send extra contrast diagnostics, for example:
```Javascript
test(`evaluate() is applicable to an <a> element with a <p> parent element
    when at least one child textnode of the <p> element has a different foreground color from the <a> element`, async (t) => {
  const target = <a href="#">Link</a>;

  const document = h.document(
    [
      <p>
        <span>
          Hello
          <span>World</span>
        </span>
        {target}
        <strong>Differently colored text</strong>
      </p>,
    ],
    [
      h.sheet([
        h.rule.style("span", {
          color: "#0000EE",
        }),
      ]),
    ]
  );

  t.deepEqual(await evaluate(ER62, { document }), [
    passed(ER62, target, {
      1: Outcomes.IsDistinguishable([defaultStyle], [hoverStyle], [focusStyle]),
    }),
  ]);
});
``` 
Pairwise contrast diagnostics are generated then for the `<a>` and `<span>` comparison [here](https://github.com/Siteimprove/alfa/blob/main/packages/alfa-rules/src/sia-er62/rule.ts#L173-L190).

Reply:
🤔 That case shouldn't be affected by this change. Both `containers` and `applicableContainers` contain the `<a>` => `<p>` mapping in that case.

Now there is an interesting question of what the rule should do for
```html
<p>
Hello <a href="…" style="text-decoration: none">link</a> <strong style="color: "#0000EE">world</strong>
</p>
```
(or similar case) The link has different color from "hello" and is therefore applicable, but it has same styling as "hello"; and different styling from "world". Should it pass or fail?

I think this is again a borderline case that we want to accept just because there are way too many corner cases to be sure this would be bad (and I am not sure this would actually be bad for WCAG…)

So, I am OK with comparing the style of the link text with all the container text, including the ones that had the same color to start with.

---
The change would be for
```html
<p>Hello <a href="…" style="color: black">link</a> </p>
<p>World <a href="…">link</a> </p>
```
where the first link is not applicable (due to same color) and thus `container` should contain both `<a>` => `<p>` mappings, but `applicableContainers` only contain the second.
However, in that case, `nonLinkElements` (in the expectation) being built from `containers.get(target)` should never contain the first `<p>`.